GoogleMulticastAppDelegate/Sources/MulticastAppDelegate.swift

```diff
@@ -0,0 +1,153 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import UIKit
+
+private typealias Application = UIApplication
+private typealias ApplicationDelegate = UIApplicationDelegate
+
+@objc(GULMulticastAppDelegateProtocol)
+public protocol MulticastAppDelegateProtocol: NSObjectProtocol {
+  typealias Delegate = UIApplicationDelegate
+
+  func addInterceptor(_ interceptor: Delegate)
+  func removeInterceptor(_ interceptor: Delegate)
+}
+
+@objc(GULMulticastAppDelegate)
+open class MulticastAppDelegate: NSObject, MulticastAppDelegateProtocol {
+  public var appDelegate: MulticastAppDelegateProtocol.Delegate?
+  private var interceptors: [MulticastAppDelegateProtocol.Delegate] = []
+  private var allInterceptors: [MulticastAppDelegateProtocol.Delegate] {
+    guard let appDelegate = appDelegate else {
+      return interceptors
+    }
+
+    var allInterceptors = [appDelegate]
+    allInterceptors.append(contentsOf: interceptors)
+    return allInterceptors
+  }
+
+  override public init() {
+    super.init()
+  }
+
+  public init(appDelegate: MulticastAppDelegateProtocol.Delegate) {
+    super.init()
+    self.appDelegate = appDelegate
+  }
+
+  @objc
+  public func addInterceptor(_ interceptor: Delegate) {
+    interceptors.append(interceptor)
+  }
+
+  @objc
+  public func removeInterceptor(_ interceptor: Delegate) {
+    interceptors = interceptors.filter { $0 !== interceptor }
+  }
+
+  // Forward all unknown messages to the original app delegate.
+  override public func responds(to aSelector: Selector!) -> Bool {
+    if type(of: self).instancesRespond(to: aSelector) {
+      return true
+    }
+
+    return appDelegate?.responds(to: aSelector) ?? false
+  }
+
+  override open func forwardingTarget(for aSelector: Selector) -> Any? {
+    return appDelegate
+  }
+}
+
+// MARK: - Multicast App Delegate detection
+
+extension MulticastAppDelegate {
+  /// Returns an instance of app delegate if it conforms to `MulticastAppDelegateProtocol`
+  @objc
+  public class func installedMulticastDelegate() -> MulticastAppDelegateProtocol? {
+    guard let appDelegate = Application.shared.delegate else {
+      return nil
+    }
+
+    if let multicastDelegate = appDelegate as? MulticastAppDelegateProtocol {
```

Comment:
This will likely never work in Swift, so the fallback below is the main code path.